src/enumerable.cr

```diff
@@ -1747,11 +1747,27 @@ module Enumerable(T)
   # ```
   # ["a", "A", "b", "B"].tally_by(&.downcase) # => {"a" => 2, "b" => 2}
   # ```
-  def tally_by(& : T -> U) : Hash(U, Int32) forall U
-    each_with_object(Hash(U, Int32).new) do |item, hash|
+  def tally_by(&block : T -> U) : Hash(U, Int32) forall U
+    tally_by(Hash(U, Int32).new, &block)
+  end
+
+  # Tallies the collection. Accepts a `hash` to count occurrences.
+  # The value corresponding to each element must be an integer.
+  # Returns a `hash` where the keys are the
+  # elements and the values are numbers of elements in the collection
+  # that correspond to the key after transformation by the given block.
+  #
+  # ```
+  # hash = {} of Char => Int32
+  # words = ["Crystal", "Ruby"]
+  # words.each { |word| word.chars.tally_by(hash, &.downcase) }
+  # hash
+  # => {'c' => 1, 'r' => 2, 'y' => 2, 's' => 1, 't' => 1, 'a' => 1, 'l' => 1, 'u' => 1, 'b' => 1}
+  # ```
+  def tally_by(hash)
+    each_with_object(hash) do |item, hash|
       value = yield item
-      count = hash[value]?
-      hash[value] = count ? count + 1 : 1
+      hash[value] = hash.fetch(value, 0) + 1
```

Comment:
I don't think this will work if the Hash is, for example, `Hash(Char, Int64)`. You can try it.

I think we need to do something like:

```crystal
      count = hash[value]?
      hash[value] = (count || typeof(count.not_nil!).zero) + 1
```

Also please add a spec for the Int64 case 🙏